File: backend/internal/models/transaction.go
package models

import "database/sql"

// TransactionType represents the kind of financial transaction.
type TransactionType string

const (
	TypeIncome  TransactionType = "INCOME"
	TypeExpense TransactionType = "EXPENSE"
)

// ExpenseStatus represents the payment status of an expense.
type ExpenseStatus string

const (
	StatusPending ExpenseStatus = "PENDING"
	StatusPaid    ExpenseStatus = "PAID"
)

// Transaction is the core domain model representing a financial transaction.
// For INCOME transactions: Status and PaidAmount are always nil/null.
// For EXPENSE transactions: Status is required (PENDING/PAID), PaidAmount >= 0.
type Transaction struct {
	ID          string          `json:"id"`
	Type        TransactionType `json:"type"`
	Category    string          `json:"category"`
	Description string          `json:"description"`
	Amount      float64         `json:"amount"`
	Date        string          `json:"date"`       // YYYY-MM-DD
	Status      *ExpenseStatus  `json:"status"`      // nil for INCOME
	PaidAmount  *float64        `json:"paidAmount"`  // nil for INCOME
	GroupID     *string         `json:"groupId"`     // Owner group
	UserID      *string         `json:"userId"`      // Creator user
	CreatedAt   string          `json:"createdAt"`
	UpdatedAt   string          `json:"updatedAt"`
}

// TransactionRow is used for scanning database rows where nullable fields
// are represented as sql.Null* types.
type TransactionRow struct {
	ID          string
	Type        string
	Category    string
	Description string
	Amount      float64
	Date        string
	Status      sql.NullString
	PaidAmount  sql.NullFloat64
	GroupID     sql.NullString
	UserID      sql.NullString
	CreatedAt   string
	UpdatedAt   string
}

// ToTransaction converts a database row representation to the domain model.
func (r *TransactionRow) ToTransaction() Transaction {
	t := Transaction{
		ID:          r.ID,
		Type:        TransactionType(r.Type),
		Category:    r.Category,
		Description: r.Description,
		Amount:      r.Amount,
		Date:        r.Date,
		CreatedAt:   r.CreatedAt,
		UpdatedAt:   r.UpdatedAt,
	}

	if r.Status.Valid {
		status := ExpenseStatus(r.Status.String)
		t.Status = &status
	}

	if r.PaidAmount.Valid {
		t.PaidAmount = &r.PaidAmount.Float64
	}

	if r.GroupID.Valid {
		t.GroupID = &r.GroupID.String
	}

	if r.UserID.Valid {
		t.UserID = &r.UserID.String
	}

	return t
}

// --- Request / Response DTOs ---

// CreateTransactionRequest is the payload for POST /api/transactions.
type CreateTransactionRequest struct {
	Type        TransactionType `json:"type"`
	Category    string          `json:"category"`
	Description string          `json:"description"`
	Amount      float64         `json:"amount"`
	Date        string          `json:"date"`
	Status      *ExpenseStatus  `json:"status,omitempty"`
	PaidAmount  *float64        `json:"paidAmount,omitempty"`
}

// UpdateTransactionRequest is the payload for PATCH /api/transactions/{id}.
type UpdateTransactionRequest struct {
	Amount     *float64       `json:"amount,omitempty"`
	Status     *ExpenseStatus `json:"status,omitempty"`
	PaidAmount *float64       `json:"paidAmount,omitempty"`
}

// TransactionSummary holds aggregated dashboard metrics for a given month.
type TransactionSummary struct {
	TotalIncome  float64 `json:"totalIncome"`
	TotalPaid    float64 `json:"totalPaid"`
	TotalPending float64 `json:"totalPending"`
}

// TransactionsResponse is the envelope for GET /api/transactions.
type TransactionsResponse struct {
	Transactions []Transaction      `json:"transactions"`
	Summary      TransactionSummary `json:"summary"`
}

// ErrorResponse is a standard error payload returned by the API.
type ErrorResponse struct {
	Error   string `json:"error"`
	Details string `json:"details,omitempty"`
}

File: frontend/src/types/transaction.ts
// =============================================================================
// Transaction Type System
// Uses TypeScript discriminated unions to enforce that:
//   - INCOME transactions never have status or paidAmount
//   - EXPENSE transactions always have status and paidAmount
// =============================================================================

/** Transaction type discriminator */
export type TransactionType = "INCOME" | "EXPENSE";

/** Expense payment status */
export type ExpenseStatus = "PENDING" | "PAID";

// --- Domain Models (match Go structs / DB schema) ---

interface BaseTransaction {
  id: string;
  category: string;
  description: string;
  amount: number;
  date: string; // YYYY-MM-DD
  userId?: string;          // wallet owner
  createdById?: string;     // who recorded this
  receiptImage?: string;
  createdByUsername?: string; // recorder username (multi-member groups)
  ownerUsername?: string;     // wallet owner username (multi-member groups)
  createdAt: string;
  updatedAt: string;
}

/** Income transaction — status & paidAmount are compile-time prohibited */
export interface IncomeTransaction extends BaseTransaction {
  type: "INCOME";
  status: null;
  paidAmount: null;
}

/** Expense transaction — status & paidAmount are required */
export interface ExpenseTransaction extends BaseTransaction {
  type: "EXPENSE";
  status: ExpenseStatus;
  paidAmount: number;
}

/** Discriminated union — narrows via `type` field */
export type Transaction = IncomeTransaction | ExpenseTransaction;

// --- API Request Types ---

export interface CreateIncomeRequest {
  type: "INCOME";
  category: string;
  description: string;
  amount: number;
  date: string;
  userId?: string;       // optional: override wallet owner
  receiptImage?: string;
}

export interface CreateExpenseRequest {
  type: "EXPENSE";
  category: string;
  description: string;
  amount: number;
  date: string;
  status: ExpenseStatus;
  paidAmount: number;
  userId?: string;       // optional: override wallet owner
  receiptImage?: string;
}

export type CreateTransactionRequest = CreateIncomeRequest | CreateExpenseRequest;

export interface UpdateTransactionRequest {
  amount?: number;
  status?: ExpenseStatus;
  paidAmount?: number;
  userId?: string;       // optional: change wallet owner
  receiptImage?: string;
}

// --- API Response Types ---

export interface TransactionSummary {
  totalIncome: number;
  totalPaid: number;
  totalPending: number;
}

export interface TransactionsResponse {
  transactions: Transaction[];
  summary: TransactionSummary;
}

/** Lightweight transaction used for month-over-month comparison (no receipt image, status, or timestamps) */
export interface CompactTransaction {
  id: string;
  type: TransactionType;
  category: string;
  description: string;
  amount: number;
  userId?: string;
}

export interface CompactTransactionsResponse {
  transactions: CompactTransaction[];
}

export interface MonthlySummary {
  month: number;
  income: number;
  expense: number;
}

export interface CategorySummary {
  category: string;
  type: TransactionType;
  amount: number;
}

export interface AnnualSummaryResponse {
  year: number;
  totalIncome: number;
  totalExpense: number;
  netBalance: number;
  monthlyData: MonthlySummary[];
  categoryData: CategorySummary[];
}

export interface ErrorResponse {
  error: string;
  details?: string;
}

// --- Form State ---

export interface TransactionFormState {
  type: TransactionType;
  category: string;
  description: string;
  amount: string; // string for controlled input binding
  date: string;
  status: ExpenseStatus;
  paidAmount: string; // string for controlled input binding
  userId?: string;    // optional: wallet owner override
  receiptImage?: string;
}

// --- Month-over-Month Change ---

export interface MomChange {
  previousAmount: number;
  currentAmount: number;
  percentChange: number; // positive = increased, negative = decreased
  isNew: boolean;        // no matching transaction in previous month
  isUnchanged: boolean;  // exact same amount
  prevTxId: string;      // previous month's transaction ID (empty if new)
}

export type MomChangeMap = Record<string, MomChange>; // keyed by current tx ID

// --- Utility type guards ---

export function isExpense(t: Transaction): t is ExpenseTransaction {
  return t.type === "EXPENSE";
}

export function isIncome(t: Transaction): t is IncomeTransaction {
  return t.type === "INCOME";
}

// --- Wallet Summary ---

export interface WalletMemberSummary {
  userId: string;
  username: string;
  totalIncome: number;
  totalExpense: number;
  totalPaid: number;
  totalPending: number;
  netBalance: number;
}

export interface WalletSummaryResponse {
  month: number;
  year: number;
  members: WalletMemberSummary[];
  groupTotal: WalletMemberSummary;
}

// --- Categories ---

export const INCOME_CATEGORIES = [
  "Salary",
  "Freelance",
  "Investment",
  "Rental",
  "Gift",
  "Refund",
  "Other",
] as const;

export const EXPENSE_CATEGORIES = [
  "Food & Dining",
  "Transportation",
  "Housing",
  "Utilities",
  "Healthcare",
  "Entertainment",
  "Shopping",
  "Education",
  "Insurance",
  "Savings",
  "Other",
] as const;
